### PR TITLE
TASK: Move object validation out of PersistenceManager

### PR DESCRIPTION
**Pull request by kdambekalns**
This moves the code used to validate objects using the onFlush event
of Doctrine out of the PersistenceManager and into it's own class.

The listener is then added using Flow settings, instead of PHP code
inside PersistenceManager.initializeObject().

### CODE REVIEW

File: Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+    /**
+     * @Flow\Inject
+     * @var ReflectionService
+     */
+    protected $reflectionService;
+
+    /**
+     * @Flow\Inject
+     * @var PersistenceManagerInterface
```
Comment (kitsunet) [inline Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php]: Might as well annotate the doctrine persistence manager implementation as it won't work with any other persistence manager anyway.

File: Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+                    unset($entityInsertions[spl_object_hash($entity)]);
+                    continue;
+                }
+
+                $knownValueObjects[$className][$identifier] = true;
+            }
+            $this->validateObject($entity, $validatedInstancesContainer);
+        }
+
+        ObjectAccess::setProperty($unitOfWork, 'entityInsertions', $entityInsertions, true);
```
Comment (kitsunet) [inline Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php]: TIL. Didn't know we did this kind of nasty magic there... But probably no other way currently?

File: Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+use Neos\Flow\Persistence\Exception\ObjectValidationFailedException;
+use Neos\Flow\Persistence\PersistenceManagerInterface;
+use Neos\Flow\Reflection\ClassSchema;
+use Neos\Utility\ObjectAccess;
+use Neos\Flow\Reflection\ReflectionService;
+use Neos\Utility\TypeHandling;
+use Neos\Flow\Validation\ValidatorResolver;
+
+/**
+ * Flow's Doctrine PersistenceManager
```
Comment (kitsunet) [inline Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php]: Not quite

**Comment by kdambekalns** [comment]
Thanks for the review… "found" this branch while doing upmerges and decided to push it into a PR. :)

File: Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php (hunk 2 above)
Comment (kdambekalns) [reply Neos.Flow/Classes/Persistence/Doctrine/ObjectValidationListener.php]: TBH I don't know. We sure did it for a reson back then…

**Comment by kitsunet** [comment]
Generally yes, I just wonder if we shouldn't clal the listener class something more insightful, eg. `ObjectValidationListener` or `ObjectValidationOnFlushListener`.

**Comment by kdambekalns** [comment]
Because it is used

> to de-duplicate value objects and validate […]

That _could_ be split in two, of course…

**Comment by kdambekalns** [comment]
THB I don't get the test failure. And I am sure it's unrelated to this PR, so… 🤷‍♂️

**Comment by kdambekalns** [comment]
_That_ should work now… 🙄

**Pull request merged**